duckduckgo/iOS
# Mac Promo Experiment 2

Task/Issue URL: https://app.asana.com/0/72649045549333/1204632216628999/f
Tech Design URL:
CC:

**Description**:

Adds support for Mac Promo Experiment 2 which shows a full screen / page sheet view and tests against the remote message framework. 

Minor changes were made to BSK (see https://github.com/duckduckgo/BrowserServicesKit/pull/375 )

**Steps to test this PR**:

***Pre-reqs:***
1. Between each test delete the app and get through onboarding. 

***Test Sheet/Takeover View***:
1. Edit `MacPromoExperiment.swift`
- Return a message from the `message` property.  The sheet doesn't actually use this but it still needs to be a "bigSingleAction" with "share" action configured message, e.g. 
```
    var message: RemoteMessageModel? {
        // return remoteMessagingStore.fetchScheduledRemoteMessage()
        return .init(id: Self.promoId, content:
                .bigSingleAction(titleText: "title",
                                 descriptionText: "description",
                                 placeholder: .macComputer,
                                 primaryActionText: "primary",
                                 primaryAction: .share(url: "https://example.com", title: "Example!")),
                     matchingRules: [], exclusionRules: [])
    }
```
- Update the property wrapper of the cohort var so the default is `Cohort.sheet.rawValue`

```
    // @UserDefaultsWrapper(key: .macPromo23Exp2, defaultValue: Cohort.unassigned.rawValue)
    private var cohortValue: String = Cohort.sheet.rawValue
```

2. Launch the app.
- Phone should show as a full screen page sheet
- iPad should show as a page sheet in centre of screen
3. Check the UI looks as expected
4. Check pixels are fired at the appropriate times
- macPromoSheetShownUnique
- macPromoPrimaryActionClicked
- macPromoSheetDismissed
- shareLink (with parameters, `success`, `cohort` and `error` if appropriate )
5. Clicking primary action should show share sheet
6. Dismissing share sheet should close the UI

***Test message view***

Note that debug message config is pointing to a version of the Mac promo that ignores d3-9 for the purpose of testing.

1. Smoke test this view - again override the message property as above.
2. Ensure that a message with a different id is always displayed as expected.
3. Ensure that a message with the mac promo id is only displayed when the message cohort is allocated.
4. Check that buttons do not dismiss the UI when the button is a share link - but should get dismissed once the share sheet is dismissed
5. Check that buttons DO dismiss the UI when the button is a url or App Store action.




**Copy Testing**:


**Orientation Testing**:


**Device Testing**:


**OS Testing**:


**Theme Testing**:


---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
